// @metaplex-foundation/umi-uploader-irys/dist/cjs/webPlugin.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var umi = require('@metaplex-foundation/umi');
var umiWeb3jsAdapters = require('@metaplex-foundation/umi-web3js-adapters');
var web3_js = require('@solana/web3.js');
var webUpload = require('@irys/web-upload');
var webUploadSolana = require('@irys/web-upload-solana');
var createIrysUploader$1 = require('./createIrysUploader.cjs');
var errors = require('./errors.cjs');

const irysUploader = options => ({
  install(umi) {
    umi.uploader = createIrysUploader(umi, options);
  }
});
function createIrysUploader(context, uploaderOptions = {}) {
  return createIrysUploader$1.createBaseIrysUploader(initWebIrys, context, uploaderOptions);
}
const initWebIrys = async (address, payer, options, context) => {
  const wallet = {
    publicKey: umiWeb3jsAdapters.toWeb3JsPublicKey(payer.publicKey),
    signMessage: message => payer.signMessage(message),
    signTransaction: async web3JsTransaction => umiWeb3jsAdapters.toWeb3JsLegacyTransaction(await payer.signTransaction(umiWeb3jsAdapters.fromWeb3JsLegacyTransaction(web3JsTransaction))),
    signAllTransactions: async web3JsTransactions => {
      const transactions = web3JsTransactions.map(umiWeb3jsAdapters.fromWeb3JsLegacyTransaction);
      const signedTransactions = await payer.signAllTransactions(transactions);
      return signedTransactions.map(umiWeb3jsAdapters.toWeb3JsLegacyTransaction);
    },
    sendTransaction: async (web3JsTransaction, connection, options = {}) => {
      const {
        signers: web3JsSigners = [],
        ...sendOptions
      } = options;
      const signers = web3JsSigners.map(web3JsSigner => umi.createSignerFromKeypair(context, umiWeb3jsAdapters.fromWeb3JsKeypair(web3_js.Keypair.fromSecretKey(web3JsSigner.secretKey))));
      let transaction = umiWeb3jsAdapters.fromWeb3JsLegacyTransaction(web3JsTransaction);
      transaction = await umi.signTransaction(transaction, [payer, ...signers]);
      const signature = await context.rpc.sendTransaction(transaction, {
        ...sendOptions,
        preflightCommitment: sendOptions.preflightCommitment
      });
      return umi.base58.deserialize(signature)[0];
    }
  };
  try {
    return await webUpload.WebUploader(webUploadSolana.WebSolana).withProvider(wallet).bundlerUrl(address).withIrysConfig(options).build();
  } catch (error) {
    throw new errors.FailedToInitializeIrysError(error);
  }
};

exports.createIrysUploader = createIrysUploader;
exports.initWebIrys = initWebIrys;
exports.irysUploader = irysUploader;
