// @metaplex-foundation/umi-uploader-irys/dist/esm/createIrysUploader.mjs
import { createGenericFileFromJson, publicKey, lamports } from '@metaplex-foundation/umi';
import BigNumber from 'bignumber.js';
import { Buffer } from 'buffer';
import { PromisePool } from '@supercharge/promise-pool';
import { IrysAbortError, AssetUploadFailedError, IrysWithdrawError, FailedToConnectToIrysAddressError } from './errors.mjs';

// PromisePool is a dependency the Irys client already requires, so using it here has no extra cost.

// Size of Irys transaction header.
const HEADER_SIZE = 2_000;

// Minimum file size for cost calculation.
const MINIMUM_SIZE = 80_000;
const gatewayUrl = id => `https://gateway.irys.xyz/${id}`;
function createBaseIrysUploader(initFn, context, uploaderOptions = {}) {
  // eslint-disable-next-line @typescript-eslint/naming-convention
  let _irys = null;
  uploaderOptions = {
    providerUrl: context.rpc.getEndpoint(),
    ...uploaderOptions
  };
  const getUploadPriceFromBytes = async bytes => {
    const irys = await getIrys();
    const price = await irys.getPrice(bytes);
    return bigNumberToAmount(price.multipliedBy(uploaderOptions.priceMultiplier ?? 1.1));
  };
  const getUploadPrice = async files => {
    const bytes = files.reduce((sum, file) => sum + HEADER_SIZE + Math.max(MINIMUM_SIZE, file.buffer.byteLength), 0);
    return getUploadPriceFromBytes(bytes);
  };
  const upload = async (files, options) => {
    const irys = await getIrys();
    const amount = await getUploadPrice(files);
    await fund(amount);
    const manifestMap = options?.manifest === true ? new Map() : undefined;
    const result = await PromisePool.for(files).withConcurrency(uploaderOptions.uploadConcurrency ?? 10).onTaskFinished((_, pool) => options?.onProgress?.(pool.processedPercentage())).process(async file => {
      if (options?.signal?.aborted) throw new IrysAbortError();
      const buffer = Buffer.from(file.buffer);
      const irysTx = irys.createTransaction(buffer, {
        tags: getGenericFileTagsWithContentType(file)
      });
      await irysTx.sign();
      const {
        status,
        data: {
          id
        }
      } = await irys.uploader.uploadTransaction(irysTx);
      if (status >= 300) throw new AssetUploadFailedError(status);
      manifestMap?.set(file.fileName, id);
      return id;
    });
    if (manifestMap) {
      const manifest = await irys.uploader.generateFolder({
        items: manifestMap
      });
      const {
        id
      } = await irys.upload(JSON.stringify(manifest), {
        tags: [{
          name: 'Type',
          value: 'manifest'
        }, {
          name: 'Content-Type',
          value: 'application/x.irys-manifest+json'
        }
        // ...(options?.manifestTags ?? []),
        ]
      });

      return [gatewayUrl(id)];
    }
    return result.results.map(gatewayUrl);
  };
  const uploadJson = async json => {
    const file = createGenericFileFromJson(json);
    const uris = await upload([file]);
    return uris[0];
  };
  const getBalance = async () => {
    const irys = await getIrys();
    const balance = await irys.getBalance();
    return bigNumberToAmount(balance);
  };
  const fund = async (amount, skipBalanceCheck = false) => {
    const irys = await getIrys();
    let toFund = amountToBigNumber(amount);
    if (!skipBalanceCheck) {
      const balance = await irys.getBalance();
      toFund = toFund.isGreaterThan(balance) ? toFund.minus(balance) : new BigNumber(0);
    }
    if (toFund.isLessThanOrEqualTo(0)) {
      return;
    }
    await irys.fund(toFund);
  };
  const withdrawAll = async () => {
    const irys = await getIrys();
    const balance = await irys.getBalance();
    const minimumBalance = new BigNumber(5000);
    if (balance.isLessThan(minimumBalance)) {
      return;
    }
    await irys.withdrawAll();
  };
  const withdraw = async amount => {
    const irys = await getIrys();
    try {
      await irys.withdrawBalance(amountToBigNumber(amount));
    } catch (e) {
      throw new IrysWithdrawError(e instanceof Error ? e.message : e.toString());
    }
  };
  const getIrys = async () => {
    const oldPayer = _irys?.getSigner().publicKey;
    const newPayer = uploaderOptions.payer ?? context.payer;
    if (oldPayer && publicKey(new Uint8Array(oldPayer)) !== newPayer.publicKey) {
      _irys = null;
    }
    if (!_irys) {
      _irys = await initIrys();
    }
    return _irys;
  };
  const initIrys = async () => {
    const token = 'solana';
    const defaultAddress = context.rpc.getCluster() === 'devnet' ? 'https://devnet.irys.xyz' : 'https://uploader.irys.xyz';
    const address = uploaderOptions?.address ?? defaultAddress;
    const irysOptions = {
      timeout: uploaderOptions.timeout,
      providerUrl: uploaderOptions.providerUrl
    };
    const payer = uploaderOptions.payer ?? context.payer;
    const irys = await initFn(address, payer, irysOptions, context);
    try {
      // Check for valid irys node.
      await irys.utils.getBundlerAddress(token);
    } catch (error) {
      throw new FailedToConnectToIrysAddressError(address, error);
    }
    return irys;
  };
  return {
    getUploadPriceFromBytes,
    getUploadPrice,
    upload,
    uploadJson,
    getBalance,
    fund,
    withdrawAll,
    withdraw,
    irys: getIrys
  };
}
const isIrysUploader = uploader => 'irys' in uploader && 'getBalance' in uploader && 'fund' in uploader && 'withdrawAll' in uploader;
const bigNumberToAmount = bigNumber => lamports(bigNumber.decimalPlaces(0).toString());
const amountToBigNumber = amount => new BigNumber(amount.basisPoints.toString());
const getGenericFileTagsWithContentType = file => {
  if (!file.contentType) {
    return file.tags;
  }
  return [{
    name: 'Content-Type',
    value: file.contentType
  }, ...file.tags];
};

export { createBaseIrysUploader, isIrysUploader };
