// @metaplex-foundation/umi-uploader-irys/dist/esm/webPlugin.mjs
import { createSignerFromKeypair, signTransaction, base58 } from '@metaplex-foundation/umi';
import { toWeb3JsPublicKey, toWeb3JsLegacyTransaction, fromWeb3JsLegacyTransaction, fromWeb3JsKeypair } from '@metaplex-foundation/umi-web3js-adapters';
import { Keypair } from '@solana/web3.js';
import { WebUploader } from '@irys/web-upload';
import { WebSolana } from '@irys/web-upload-solana';
import { createBaseIrysUploader } from './createIrysUploader.mjs';
import { FailedToInitializeIrysError } from './errors.mjs';

const irysUploader = options => ({
  install(umi) {
    umi.uploader = createIrysUploader(umi, options);
  }
});
function createIrysUploader(context, uploaderOptions = {}) {
  return createBaseIrysUploader(initWebIrys, context, uploaderOptions);
}
const initWebIrys = async (address, payer, options, context) => {
  const wallet = {
    publicKey: toWeb3JsPublicKey(payer.publicKey),
    signMessage: message => payer.signMessage(message),
    signTransaction: async web3JsTransaction => toWeb3JsLegacyTransaction(await payer.signTransaction(fromWeb3JsLegacyTransaction(web3JsTransaction))),
    signAllTransactions: async web3JsTransactions => {
      const transactions = web3JsTransactions.map(fromWeb3JsLegacyTransaction);
      const signedTransactions = await payer.signAllTransactions(transactions);
      return signedTransactions.map(toWeb3JsLegacyTransaction);
    },
    sendTransaction: async (web3JsTransaction, connection, options = {}) => {
      const {
        signers: web3JsSigners = [],
        ...sendOptions
      } = options;
      const signers = web3JsSigners.map(web3JsSigner => createSignerFromKeypair(context, fromWeb3JsKeypair(Keypair.fromSecretKey(web3JsSigner.secretKey))));
      let transaction = fromWeb3JsLegacyTransaction(web3JsTransaction);
      transaction = await signTransaction(transaction, [payer, ...signers]);
      const signature = await context.rpc.sendTransaction(transaction, {
        ...sendOptions,
        preflightCommitment: sendOptions.preflightCommitment
      });
      return base58.deserialize(signature)[0];
    }
  };
  try {
    return await WebUploader(WebSolana).withProvider(wallet).bundlerUrl(address).withIrysConfig(options).build();
  } catch (error) {
    throw new FailedToInitializeIrysError(error);
  }
};

export { createIrysUploader, initWebIrys, irysUploader };
